### DISCUSSION TITLE
Typing ESC in GitBash/Git4Windows temporarily freezes keyboard input in that Windows Terminal tab (except for CTRL-C), Windows 10

### DISCUSSION THREAD

**Discussion by coderextreme** [microsoft/terminal#17252] [Q&A]
### Windows Terminal version

1.19.11213.0

### Windows build number

10.0.19045.0

### Other Software

$ uname -a
MINGW64_NT-10.0-19045 DESKTOP-8OPL3IE 3.4.10-87d57229.x86_64 2024-02-14 20:17 UTC x86_64 Msys
$ for n in 0 1 2 3 4 5
do
echo "BASH_VERSINFO[$n] = ${BASH_VERSINFO[$n]}"
done
BASH_VERSINFO[0] = 5
BASH_VERSINFO[1] = 2
BASH_VERSINFO[2] = 26
BASH_VERSINFO[3] = 1
BASH_VERSINFO[4] = release
BASH_VERSINFO[5] = x86_64-pc-msys

Git 4 Windows 2.45.0 and lower.

### Steps to reproduce

Install Git4Windows/GitBash. Type ESC (perhaps several times) at the Bash command prompt in Windows Terminal, after selecting Git Bash and then type space or another keystroke.  The terminal will not respond (freeze) for a while, then continue.  This also affects the MingGW app under Windows and WSLg/WSL2 under Ubuntu 24.  I have not noticed it inside other console terminals or other shells, such as PowerShell or CommandPrompt.

### Expected Behavior

I was expecting an immediate response after typing ESC, or the ESC key should perhaps be echoed while typing at the bash command-line.  I realize that ESC may start a terminal sequence, so it may be desirable not to echo it out.  I use vim a lot, so typing ESC is very common for me.  I recently figured out that typing ESC at the Bash command prompt was the issue after cursing you guys for a long time.  No other terminal I've used (except for MingGW today) has ever had this issue.

### Actual Behavior

The terminal temporarily freezes when you type ESC.  If there was a way to cancel the freeze, or unlock the terminal with another keystroke, that would be satisfactory.
Please no comments on my use of vim, it's unrelated at this point, and pointing at the user for their behavior is uncool.  There's a joke about a patient visiting that doctor, complaining of a pain when the patient does something, and the doctor says: "Don't do that!" as a response. The behavior in Command Prompt or PowerShell, cancelling the spaces typed, and returning the cursor to "Home" would be satisfactory, as I usually type ESC at the start of a line.

**Comment by coderextreme** [comment]
Upgrading, ```pacman -Suy``` solved my issue.  I don't know why installing new versions of GitBash/GitForWindows did not do this.

**Comment by coderextreme** [comment]
Actually, if I type several spaces, then ESC, my bash freezes.  I want the behavior of Command Prompt/PowerShell, if possible.

**Comment by coderextreme** [comment]
The error has returned in Windows Terminal and Git Bash, but MINGW64 and Msys2 terminals work ok.

**Comment by varverious** [comment]
Full answer is at https://superuser.com/questions/1930161/why-does-git-bash-hang-for-several-seconds-if-i-press-tab-as-the-first-key-and
Here's a gist of it:

It's trying to autocomplete nothing, which by default resolves to every possibility in Bash (Git Bash is just a Windows version of it), including all the executables from all the %PATH% directories, etc.
You want to prevent autocompletion of empty input. This is done by e.g. enabling the no_empty_cmd_completion [shell option](https://www.gnu.org/software/bash/manual/html_node/The-Shopt-Builtin.html):

`shopt -s no_empty_cmd_completion`

It disables it for both completely empty and whitespace-only lines. You can add it to [.bashrc](https://stackoverflow.com/questions/6883760/git-for-windows-bashrc-or-equivalent-configuration-files-for-git-bash-shell) file for it to be permament.
